### PR TITLE
Link parser rewrite

### PR DESCRIPTION
**Pull request by Umcaruje**
Here is a rewrite of the parser function, so we don't need to use all sorts of hacks to make it work :)

I'd like a second set of eyes to look at the code (@Lukas-W or @tresf ) before I merge this, just in case.

### CODE REVIEW

File: public/lsp/utils.php [hunk 1]
```diff
@@ -383,25 +383,59 @@ function get_file_url($file_id = null) {
...
+					preg_match("#\s*[a-zA-Z\/\/:\.]*soundcloud.com\/([a-zA-Z0-9\*\-\_\?\&\;\%\=\.]+)\/([a-zA-Z0-9\*\-\_\?\&\;\%\=\.]+)\/*([a-zA-Z0-9\*\-\_\?\&\;\%\=\.]*)#i", $matched[$i][0], $sc);
+					if (! empty($sc)) {
+						// Do not process playlists
+						if (! $sc[3]) {
+						$message = str_replace($sc[0], "<sc>$sc[1]/$sc[2]</sc>", $message);
+						$message = soundcloud_iframe($message, $width, $height);
+						} else {
+							// Check for links
+							preg_match("#((https?://)([a-zA-Z0-9\/\*\-\_\?\&\;\%\=\.])+)#i", $matched[$i][0], $link);
+							if (! empty($link)) {
```
Comment (tresf) [inline public/lsp/utils.php]: > "If you need more than 3 levels of indentation, you're screwed anyway, and should fix your program." - *Linus Torvalds* per [wikiquote.org](http://en.wikiquote.org/wiki/Linus_Torvalds)

Comment (Umcaruje) [inline public/lsp/utils.php]: True, looking at the code, its really messy and inefficient, I will try to simplify it.

**Comment by Umcaruje** [comment]
Well, I've rewritten it so regular expressions are only used once. I think its much cleaner now. @tresf thoughts?

**Comment by tresf** [comment]
> Well, I've rewritten it so regular expressions are only used once. I think its much cleaner now. @tresf thoughts?

Well, don't think that a longer equation (in any language) makes anything easier to read a year from now... ...although that is pretty cool what you've done.

When you have:

```php
if (
    if (
        if (
            if (
               if (
```

It generally means something is wrong.  That was Linus' point and I don't think that has been improved much here. :smiling_imp: 

That said, I don't mind merging this if it works... I really think the `AutoEmbed`/`MediaEmbed` type solution would bring us into a better position.

**Comment by Umcaruje** [comment]
>I really think the AutoEmbed/MediaEmbed type solution would bring us into a better position.

Ok well, first of all, both `AutoEmbed` and `MediaEmbed` do not support Soundcloud, which is the trickiest to embed, unlike youtube which is a breeze. They also support a lot of other sites, but the LSP users are pretty much using only Youtube and Soundcloud. Not to mention `AutoEmbed` is obsolete and `MediaEmbed` is in alpha.
____
The code is even simpler, since I introduced a function called `parse_url_ext` which supports links without any protocol like `soundcloud.com/user/sound`. 

>It generally means something is wrong. That was Linus' point and I don't think that has been improved much here. :smiling_imp:

I have stress tested the code, and it worked in any real case I imagined.

Also the only part that has many if statements is the soundcloud one, simply because we have to avoid embeding of playlists and old iframes.

**Comment by Umcaruje** [comment]
Well, the regex is now much more understandable to anyone, since I commented it out: 
hehehu

@tresf If you don't pose any objections in the next couple of days, I'll merge this.

**Comment by tresf** [comment]
thanks.  FYI - the tab spacing looks to be a bit weird in github.

**Comment by Umcaruje** [comment]
>FYI - the tab spacing looks to be a bit weird in github.

Fixed, merging :)

**Pull request merged**